File: mobile-sdks/flutter/lib/src/client/nigeria_geo_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:logger/logger.dart';

import '../models/state.dart';
import '../models/lga.dart';
import '../models/ward.dart';
import '../models/postal_code.dart';
import '../models/search_result.dart';
import '../models/address_validation.dart';
import '../models/pagination.dart';
import '../exceptions/nigeria_geo_exception.dart';
import '../services/cache_service.dart';
import '../utils/logger.dart';
import 'nigeria_geo_config.dart';

class NigeriaGeoClient {
  final NigeriaGeoConfig _config;
  final http.Client _httpClient;
  final CacheService _cacheService;
  final Logger _logger;

  NigeriaGeoClient._(
    this._config,
    this._httpClient,
    this._cacheService,
  ) : _logger = NigeriaGeoLogger.instance;

  /// Internal factory constructor
  static NigeriaGeoClient create(
    NigeriaGeoConfig config,
    http.Client httpClient,
    CacheService cacheService,
  ) {
    return NigeriaGeoClient._(config, httpClient, cacheService);
  }

  // States API
  Future<PaginatedResponse<NigerianState>> getStates({
    int page = 1,
    int limit = 20,
    bool useCache = true,
  }) async {
    final cacheKey = 'states_${page}_$limit';

    if (useCache && _config.enableCaching) {
      final cached = await _cacheService.get<PaginatedResponse<NigerianState>>(
        cacheKey,
        (json) => PaginatedResponse.fromJson(
          json,
          (item) => NigerianState.fromJson(item as Map<String, dynamic>),
        ),
      );
      if (cached != null) {
        _logger.d('Returning cached states for page $page');
        return cached;
      }
    }

    try {
      final response = await _makeRequest(
        'GET',
        '/api/v1/states',
        queryParameters: {'page': page.toString(), 'limit': limit.toString()},
      );

      final result = PaginatedResponse<NigerianState>.fromJson(
        response,
        (item) => NigerianState.fromJson(item as Map<String, dynamic>),
      );

      if (_config.enableCaching) {
        await _cacheService.set(
            cacheKey, result.toJson((state) => state.toJson()));
      }

      return result;
    } catch (e) {
      _logger.e('Error fetching states: $e');
      rethrow;
    }
  }

  Future<NigerianState> getStateById(String stateId) async {
    final cacheKey = 'state_$stateId';

    if (_config.enableCaching) {
      final cached = await _cacheService.get<NigerianState>(
        cacheKey,
        (json) => NigerianState.fromJson(json),
      );
      if (cached != null) return cached;
    }

    try {
      final response = await _makeRequest('GET', '/api/v1/states/$stateId');
      final state = NigerianState.fromJson(response['data']);

      if (_config.enableCaching) {
        await _cacheService.set(cacheKey, state.toJson());
      }

      return state;
    } catch (e) {
      _logger.e('Error fetching state $stateId: $e');
      rethrow;
    }
  }

  // LGAs API
  Future<PaginatedResponse<LGA>> getLGAs({
    String? stateId,
    int page = 1,
    int limit = 20,
    bool useCache = true,
  }) async {
    final cacheKey = 'lgas_${stateId ?? 'all'}_${page}_$limit';

    if (useCache && _config.enableCaching) {
      final cached = await _cacheService.get<PaginatedResponse<LGA>>(
        cacheKey,
        (json) => PaginatedResponse.fromJson(
          json,
          (item) => LGA.fromJson(item as Map<String, dynamic>),
        ),
      );
      if (cached != null) return cached;
    }

    try {
      final queryParams = <String, String>{
        'page': page.toString(),
        'limit': limit.toString(),
      };

      final String endpoint;
      if (stateId != null) {
        endpoint = '/api/v1/states/$stateId/lgas';
      } else {
        throw ArgumentError('stateId is required for fetching LGAs');
      }

      final response =
          await _makeRequest('GET', endpoint, queryParameters: queryParams);

      final result = PaginatedResponse<LGA>.fromJson(
        response,
        (item) => LGA.fromJson(item as Map<String, dynamic>),
      );

      if (_config.enableCaching) {
        await _cacheService.set(cacheKey, result.toJson((lga) => lga.toJson()));
      }

      return result;
    } catch (e) {
      _logger.e('Error fetching LGAs: $e');
      rethrow;
    }
  }

  Future<LGA> getLGAById(String lgaId) async {
    final cacheKey = 'lga_$lgaId';

    if (_config.enableCaching) {
      final cached = await _cacheService.get<LGA>(
        cacheKey,
        (json) => LGA.fromJson(json),
      );
      if (cached != null) return cached;
    }

    try {
      final response = await _makeRequest('GET', '/api/v1/lgas/$lgaId');
      final lga = LGA.fromJson(response['data']);

      if (_config.enableCaching) {
        await _cacheService.set(cacheKey, lga.toJson());
      }

      return lga;
    } catch (e) {
      _logger.e('Error fetching LGA $lgaId: $e');
      rethrow;
    }
  }

  // Wards API
  Future<PaginatedResponse<Ward>> getWards({
    String? lgaId,
    int page = 1,
    int limit = 20,
    bool useCache = true,
  }) async {
    final cacheKey = 'wards_${lgaId ?? 'all'}_${page}_$limit';

    if (useCache && _config.enableCaching) {
      final cached = await _cacheService.get<PaginatedResponse<Ward>>(
        cacheKey,
        (json) => PaginatedResponse.fromJson(
          json,
          (item) => Ward.fromJson(item as Map<String, dynamic>),
        ),
      );
      if (cached != null) return cached;
    }

    try {
      final queryParams = <String, String>{
        'page': page.toString(),
        'limit': limit.toString(),
      };

      final String endpoint;
      if (lgaId != null) {
        endpoint = '/api/v1/lgas/$lgaId/wards';
      } else {
        throw ArgumentError('lgaId is required for fetching wards');
      }

      final response =
          await _makeRequest('GET', endpoint, queryParameters: queryParams);

      final result = PaginatedResponse<Ward>.fromJson(
        response,
        (item) => Ward.fromJson(item as Map<String, dynamic>),
      );

      if (_config.enableCaching) {
        await _cacheService.set(
            cacheKey, result.toJson((ward) => ward.toJson()));
      }

      return result;
    } catch (e) {
      _logger.e('Error fetching wards: $e');
      rethrow;
    }
  }

  // Postal Codes API
  Future<PaginatedResponse<PostalCode>> getPostalCodes({
    String? wardId,
    int page = 1,
    int limit = 20,
    bool useCache = true,
  }) async {
    final cacheKey = 'postal_codes_${wardId ?? 'all'}_${page}_$limit';

    if (useCache && _config.enableCaching) {
      final cached = await _cacheService.get<PaginatedResponse<PostalCode>>(
        cacheKey,
        (json) => PaginatedResponse.fromJson(
          json,
          (item) => PostalCode.fromJson(item as Map<String, dynamic>),
        ),
      );
      if (cached != null) return cached;
    }

    try {
      final queryParams = <String, String>{
        'page': page.toString(),
        'limit': limit.toString(),
      };

      final String endpoint;
      if (wardId != null) {
        endpoint = '/api/v1/wards/$wardId/postal-codes';
      } else {
        throw ArgumentError('wardId is required for fetching postal codes');
      }

      final response =
          await _makeRequest('GET', endpoint, queryParameters: queryParams);

      final result = PaginatedResponse<PostalCode>.fromJson(
        response,
        (item) => PostalCode.fromJson(item as Map<String, dynamic>),
      );

      if (_config.enableCaching) {
        await _cacheService.set(
            cacheKey, result.toJson((postalCode) => postalCode.toJson()));
      }

      return result;
    } catch (e) {
      _logger.e('Error fetching postal codes: $e');
      rethrow;
    }
  }

  // Search API
  Future<SearchResult> searchAll(String query, {bool useCache = true}) async {
    if (query.trim().isEmpty) {
      return SearchResult(states: [], lgas: [], wards: [], postalCodes: []);
    }

    final cacheKey = 'search_${query.toLowerCase().trim()}';

    if (useCache && _config.enableCaching) {
      final cached = await _cacheService.get<SearchResult>(
        cacheKey,
        (json) => SearchResult.fromJson(json),
      );
      if (cached != null) return cached;
    }

    try {
      final response = await _makeRequest(
        'GET',
        '/api/v1/search',
        queryParameters: {'query': query.trim()},
      );

      final result = SearchResult.fromJson(response);

      if (_config.enableCaching) {
        await _cacheService.set(cacheKey, result.toJson());
      }

      return result;
    } catch (e) {
      _logger.e('Error searching for "$query": $e');
      rethrow;
    }
  }

  // Address Validation API
  Future<AddressValidationResponse> validateAddress(
    AddressValidationRequest request,
  ) async {
    try {
      final response = await _makeRequest(
        'POST',
        '/api/v1/validate',
        body: request.toJson(),
      );

      return AddressValidationResponse.fromJson(response['data']);
    } catch (e) {
      _logger.e('Error validating address: $e');
      rethrow;
    }
  }

  // Location-based queries
  Future<List<PostalCode>> findNearbyPostalCodes({
    required double latitude,
    required double longitude,
    double radiusKm = 5.0,
    int limit = 20,
  }) async {
    try {
      final response = await _makeRequest(
        'GET',
        '/api/v1/postal-codes/nearby',
        queryParameters: {
          'lat': latitude.toString(),
          'lng': longitude.toString(),
          'radius': radiusKm.toString(),
          'limit': limit.toString(),
        },
      );

      final List<dynamic> items = response['data'] ?? [];
      return items.map((item) => PostalCode.fromJson(item)).toList();
    } catch (e) {
      _logger.e('Error finding nearby postal codes: $e');
      rethrow;
    }
  }

  // Stream APIs
  Stream<List<NigerianState>> getStatesStream() {
    return Stream.fromFuture(getStates().then((response) => response.data));
  }

  Stream<List<LGA>> getLGAsStream({String? stateId}) {
    return Stream.fromFuture(
        getLGAs(stateId: stateId).then((response) => response.data));
  }

  Stream<List<Ward>> getWardsStream({String? lgaId}) {
    return Stream.fromFuture(
        getWards(lgaId: lgaId).then((response) => response.data));
  }

  // Private helper methods
  Future<Map<String, dynamic>> _makeRequest(
    String method,
    String endpoint, {
    Map<String, String>? queryParameters,
    Map<String, dynamic>? body,
    Map<String, String>? headers,
  }) async {
    final uri = Uri.parse('${_config.baseUrl}$endpoint')
        .replace(queryParameters: queryParameters);

    final requestHeaders = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      ...?headers,
    };

    if (_config.apiKey != null && _config.apiKey!.isNotEmpty) {
      requestHeaders['Authorization'] = 'Bearer ${_config.apiKey}';
    }

    if (_config.enableLogging) {
      _logger.d('$method $uri');
      if (body != null) _logger.d('Body: $body');
    }

    try {
      late http.Response response;

      switch (method.toUpperCase()) {
        case 'GET':
          response = await _httpClient
              .get(uri, headers: requestHeaders)
              .timeout(_config.timeout);
          break;
        case 'POST':
          response = await _httpClient
              .post(uri,
                  headers: requestHeaders,
                  body: body != null ? jsonEncode(body) : null)
              .timeout(_config.timeout);
          break;
        case 'PUT':
          response = await _httpClient
              .put(uri,
                  headers: requestHeaders,
                  body: body != null ? jsonEncode(body) : null)
              .timeout(_config.timeout);
          break;
        case 'DELETE':
          response = await _httpClient
              .delete(uri, headers: requestHeaders)
              .timeout(_config.timeout);
          break;
        default:
          throw UnsupportedError('HTTP method $method not supported');
      }

      if (_config.enableLogging) {
        _logger.d('Response: ${response.statusCode} ${response.body}');
      }

      return _handleResponse(response);
    } on TimeoutException {
      throw NetworkException(
          'Request timeout after ${_config.timeout.inSeconds}s');
    } on SocketException catch (e) {
      throw NetworkException('Network error: ${e.message}');
    } on http.ClientException catch (e) {
      throw NetworkException('HTTP client error: ${e.message}');
    } catch (e) {
      _logger.e('Unexpected error in _makeRequest: $e');
      rethrow;
    }
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    final statusCode = response.statusCode;

    try {
      final jsonResponse = jsonDecode(response.body) as Map<String, dynamic>;

      if (statusCode >= 200 && statusCode < 300) {
        return jsonResponse;
      } else {
        final message =
            jsonResponse['message'] ?? jsonResponse['error'] ?? 'Unknown error';

        switch (statusCode) {
          case 400:
            throw InvalidRequestException(message);
          case 401:
            throw UnauthorizedException(message);
          case 403:
            throw ForbiddenException(message);
          case 404:
            throw NotFoundException(message);
          case 429:
            throw RateLimitException(message);
          case 500:
            throw ServerException(message);
          default:
            throw ServerException('HTTP $statusCode: $message');
        }
      }
    } catch (FormatException) {
      throw InvalidResponseException('Invalid JSON response');
    }
  }

  void dispose() {
    _httpClient.close();
  }
}

File: mobile-sdks/ios/NigeriaGeoSDK/Sources/NigeriaGeoSDK/NigeriaGeoClient.swift
import Foundation

/// Main SDK client for Nigeria Geo API
@available(iOS 12.0, macOS 10.15, tvOS 12.0, watchOS 6.0, *)
public class NigeriaGeoClient {
    
    // MARK: - Properties
    
    private let config: NigeriaGeoConfig
    private let networkService: NetworkService
    private let cacheManager: CacheManager
    
    // MARK: - Initialization
    
    /// Initialize the Nigeria Geo client with configuration
    /// - Parameter config: SDK configuration
    public init(config: NigeriaGeoConfig) {
        self.config = config
        self.networkService = NetworkService(config: config)
        self.cacheManager = CacheManager(config: config)
    }
    
    /// Convenience initializer with base URL
    /// - Parameters:
    ///   - baseURL: API base URL
    ///   - apiKey: Optional API key for authentication
    public convenience init(baseURL: String, apiKey: String? = nil) {
        let config = NigeriaGeoConfig(baseURL: baseURL, apiKey: apiKey)
        self.init(config: config)
    }
}

// MARK: - States API

@available(iOS 12.0, macOS 10.15, tvOS 12.0, watchOS 6.0, *)
public extension NigeriaGeoClient {
    
    /// Get all states with pagination
    /// - Parameters:
    ///   - page: Page number (default: 1)
    ///   - limit: Items per page (default: 20)
    ///   - completion: Completion handler
    func getStates(
        page: Int = 1,
        limit: Int = 20,
        completion: @escaping (Result<PaginatedResponse<State>, NigeriaGeoError>) -> Void
    ) {
        let parameters: [String: Any] = [
            "page": page,
            "limit": limit
        ]
        
        let endpoint = APIEndpoint.states(parameters: parameters)
        
        // Check cache first
        if config.enableCaching {
            let cacheKey = endpoint.cacheKey
            if let cachedData: PaginatedResponse<State> = cacheManager.get(key: cacheKey) {
                DispatchQueue.main.async {
                    completion(.success(cachedData))
                }
                return
            }
        }
        
        // Make network request
        networkService.request(endpoint: endpoint) { [weak self] result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(PaginatedResponse<State>.self, from: data)
                    
                    // Cache the response
                    if self?.config.enableCaching == true {
                        self?.cacheManager.set(key: endpoint.cacheKey, value: response)
                    }
                    
                    DispatchQueue.main.async {
                        completion(.success(response))
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(.failure(.invalidResponse))
                    }
                }
            case .failure(let error):
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
    
    /// Get state by ID
    /// - Parameters:
    ///   - id: State UUID
    ///   - completion: Completion handler
    func getState(
        id: UUID,
        completion: @escaping (Result<State?, NigeriaGeoError>) -> Void
    ) {
        let endpoint = APIEndpoint.stateById(id: id)
        
        // Check cache first
        if config.enableCaching {
            let cacheKey = endpoint.cacheKey
            if let cachedData: State? = cacheManager.get(key: cacheKey) {
                DispatchQueue.main.async {
                    completion(.success(cachedData))
                }
                return
            }
        }
        
        networkService.request(endpoint: endpoint) { [weak self] result in
            switch result {
            case .success(let data):
                do {
                    let state = try JSONDecoder().decode(State?.self, from: data)
                    
                    // Cache the response
                    if self?.config.enableCaching == true {
                        self?.cacheManager.set(key: endpoint.cacheKey, value: state)
                    }
                    
                    DispatchQueue.main.async {
                        completion(.success(state))
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(.failure(.invalidResponse))
                    }
                }
            case .failure(let error):
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
    
    /// Get LGAs for a specific state
    /// - Parameters:
    ///   - stateId: State UUID
    ///   - page: Page number (default: 1)
    ///   - limit: Items per page (default: 20)
    ///   - completion: Completion handler
    func getLGAs(
        forState stateId: UUID,
        page: Int = 1,
        limit: Int = 20,
        completion: @escaping (Result<PaginatedResponse<LGA>, NigeriaGeoError>) -> Void
    ) {
        let parameters: [String: Any] = [
            "page": page,
            "limit": limit
        ]
        
        let endpoint = APIEndpoint.lgasByState(stateId: stateId, parameters: parameters)
        
        // Check cache first
        if config.enableCaching {
            let cacheKey = endpoint.cacheKey
            if let cachedData: PaginatedResponse<LGA> = cacheManager.get(key: cacheKey) {
                DispatchQueue.main.async {
                    completion(.success(cachedData))
                }
                return
            }
        }
        
        networkService.request(endpoint: endpoint) { [weak self] result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(PaginatedResponse<LGA>.self, from: data)
                    
                    // Cache the response
                    if self?.config.enableCaching == true {
                        self?.cacheManager.set(key: endpoint.cacheKey, value: response)
                    }
                    
                    DispatchQueue.main.async {
                        completion(.success(response))
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(.failure(.invalidResponse))
                    }
                }
            case .failure(let error):
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
}

// MARK: - Search API

@available(iOS 12.0, macOS 10.15, tvOS 12.0, watchOS 6.0, *)
public extension NigeriaGeoClient {
    
    /// Search across all geographic entities
    /// - Parameters:
    ///   - query: Search query
    ///   - page: Page number (default: 1)
    ///   - limit: Items per page (default: 20)
    ///   - completion: Completion handler
    func searchAll(
        query: String,
        page: Int = 1,
        limit: Int = 20,
        completion: @escaping (Result<SearchResult, NigeriaGeoError>) -> Void
    ) {
        let parameters: [String: Any] = [
            "query": query,
            "page": page,
            "limit": limit
        ]
        
        let endpoint = APIEndpoint.searchAll(parameters: parameters)
        
        networkService.request(endpoint: endpoint) { result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(SearchResult.self, from: data)
                    DispatchQueue.main.async {
                        completion(.success(response))
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(.failure(.invalidResponse))
                    }
                }
            case .failure(let error):
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
    
    /// Find nearby postal codes based on coordinates
    /// - Parameters:
    ///   - latitude: Latitude coordinate
    ///   - longitude: Longitude coordinate
    ///   - radiusKm: Search radius in kilometers (default: 10)
    ///   - completion: Completion handler
    func findNearbyPostalCodes(
        latitude: Double,
        longitude: Double,
        radiusKm: Double = 10.0,
        completion: @escaping (Result<[PostalCode], NigeriaGeoError>) -> Void
    ) {
        let parameters: [String: Any] = [
            "lat": latitude,
            "lng": longitude,
            "radius_km": radiusKm
        ]
        
        let endpoint = APIEndpoint.nearbyPostalCodes(parameters: parameters)
        
        networkService.request(endpoint: endpoint) { result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode([PostalCode].self, from: data)
                    DispatchQueue.main.async {
                        completion(.success(response))
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(.failure(.invalidResponse))
                    }
                }
            case .failure(let error):
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
}

// MARK: - Address Validation API

@available(iOS 12.0, macOS 10.15, tvOS 12.0, watchOS 6.0, *)
public extension NigeriaGeoClient {
    
    /// Validate an address
    /// - Parameters:
    ///   - request: Address validation request
    ///   - completion: Completion handler
    func validateAddress(
        _ request: AddressValidationRequest,
        completion: @escaping (Result<AddressValidationResponse, NigeriaGeoError>) -> Void
    ) {
        let endpoint = APIEndpoint.validateAddress
        
        networkService.request(endpoint: endpoint, body: request) { result in
            switch result {
            case .success(let data):
                do {
                    let response = try JSONDecoder().decode(AddressValidationResponse.self, from: data)
                    DispatchQueue.main.async {
                        completion(.success(response))
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(.failure(.invalidResponse))
                    }
                }
            case .failure(let error):
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
    }
}

// MARK: - Cache Management

@available(iOS 12.0, macOS 10.15, tvOS 12.0, watchOS 6.0, *)
public extension NigeriaGeoClient {
    
    /// Clear all cached data
    /// - Parameter completion: Completion handler
    func clearCache(completion: @escaping (Bool) -> Void) {
        cacheManager.clearAll()
        DispatchQueue.main.async {
            completion(true)
        }
    }
    
    /// Get cache statistics
    /// - Returns: Cache statistics
    func getCacheStatistics() -> CacheStatistics {
        return cacheManager.getStatistics()
    }
}

// MARK: - Async/Await Support (iOS 13+)

@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
public extension NigeriaGeoClient {
    
    /// Get all states with pagination (async/await)
    func getStates(page: Int = 1, limit: Int = 20) async throws -> PaginatedResponse<State> {
        return try await withCheckedThrowingContinuation { continuation in
            getStates(page: page, limit: limit) { result in
                continuation.resume(with: result)
            }
        }
    }
    
    /// Get state by ID (async/await)
    func getState(id: UUID) async throws -> State? {
        return try await withCheckedThrowingContinuation { continuation in
            getState(id: id) { result in
                continuation.resume(with: result)
            }
        }
    }
    
    /// Search across all geographic entities (async/await)
    func searchAll(query: String, page: Int = 1, limit: Int = 20) async throws -> SearchResult {
        return try await withCheckedThrowingContinuation { continuation in
            searchAll(query: query, page: page, limit: limit) { result in
                continuation.resume(with: result)
            }
        }
    }
    
    /// Validate an address (async/await)
    func validateAddress(_ request: AddressValidationRequest) async throws -> AddressValidationResponse {
        return try await withCheckedThrowingContinuation { continuation in
            validateAddress(request) { result in
                continuation.resume(with: result)
            }
        }
    }
}